[WIP] Physics behavior editor

Lizard-13 opened the pull request:
Also adds custom polygon support.

@4ian Now the parser works fine, I can make a fast test and check that un/serializing an isolated behavior content string convert the string to JSON and to string again correctly, arrays included. But for some reason, it doesn't work when opening a project file, I'm doing the following:
* Create a new project
* Add the physics behavior, the default values for `vertices` is `[]`
* Save the project >> The array is correctly saved in the game file
* Open the project >> The array is automatically converted to an empty object `{}`

Again, the parser works fine with the same isolated behavior string, am I going crazy?

4ian commented on newIDE/app/src/PropertiesEditor/PropertiesMapToSchema.js:
```diff
@@ -27,11 +28,15 @@ export default (
       .toJSArray()
       .map(value => ({ value, label: value }));
+    const dialog = PropertiesRenderingService.getDialogComponent(
+      property.getExtraInfo().toJSArray()[0]
```
This looks risky to me. [0] could return undefined and you typed `getDialogComponent` as taking a string.

Lizard-13 replied:
Ah, yeah, I know, I feel a bit bad each time I read that "line", very WIP maybe?
I've to add some extra safe checks.

4ian replied:
Maybe you could actually move this into the `getDialog` function? Would defer the call to getDialogComponent until it's really needed - so no useless call and easier to debug calls to getDialogComponent too

4ian commented:
You're 100% sure you're running the version of libGD.js with the fix right?

4ian commented:
Mmm no sorry it was another issue (forgot to load extensions in the script that create the examples files for the web-app).

Lizard-13 commented:
What would be the best way to show an image preview in the behaviors editor? passing the resources manager information to the `BehaviorsEditor` here:
https://github.com/4ian/GDevelop/blob/0cecf685ca3e7e46b4c2281c72f048dfc27bd0fe/newIDE/app/src/ObjectEditor/ObjectEditorDialog.js#L106-L113
Sounds fine, it's done for the object properties some lines before, but then passing the resource to the `PropertiesEditor` looks weird:
https://github.com/4ian/GDevelop/blob/0cecf685ca3e7e46b4c2281c72f048dfc27bd0fe/newIDE/app/src/BehaviorsEditor/index.js#L182-L185
What do you think?

4ian commented:
Sorry for the delay, mmm I guess we can start by passing resourceSources down to the PropertiesEditor. Make it an optional Prop (so that it's not mandatory for other property editors).  Will see if there is a better way, but I don't think so (when thinking about it, if a PropertiesEditor has to shown fields that need the resources... then it needs to have the resources).

Lizard-13 commented:
Now it previews almost every shape:
![shapepreview](https://user-images.githubusercontent.com/9782647/50523703-5ea67580-0ab0-11e9-927d-a9d7871b46ab.png)
The polygon editor/preview is still missing. Unless I'm forgetting something, this and investigate the bad vertices parsing is the only and last step to do.

4ian commented on newIDE/app/src/PropertiesEditor/index.js:
```diff
@@ -9,16 +9,27 @@ import MenuItem from 'material-ui/MenuItem';
   instances: Instances,
   schema: Schema,
   mode?: 'column' | 'row',
+  project?: Project,
```
use `gdProject` :) Typing for objects of libGD.js is done inside `flow-typed/libGD.js` file. Most of the definitions right now are very simple, but it's important to always use the proper type, because in the future we might want to write them properly or autogenerate them from C++ classes.

Lizard-13 commented on newIDE/app/src/PropertiesEditor/index.js:
```diff
@@ -225,9 +236,38 @@ export default class PropertiesEditor extends React.Component<Props, {||}> {
 
+    for (var i = 0, len = this.props.schema.length; i < len; i++) {
+      if (this.props.schema[i].getDialog) {
+        return (
```
But how to make it to not render fields defined in the custom dialog?
For example if I render everything it will render my dialog and the `density` property, and I need the density property defined to use it in the dialog... or an I missing something?

4ian replied:
Let me think about it, but I think we can move toward a solution that is more similar to `ObjectsEditorService`, but for Behaviors (`BehaviorsEditorService`). i.e: we register the editor that you created for the behavior. If a behavior does not have a registered editor, then `BehaviorsEditorService` will create a PropertiesEditor, like what we do now.

I was thinking of doing the same for `ObjectsEditorService`: if you don't set an editor for your object, a default editor showing the properties will be displayed.
So both `ObjectsEditorService` and `BehaviorsEditorService` would work the same. It would avoid having a strange handling of "getDialog" in PropertiesEditor, allow custom editors while providing a default editor for most behaviors/objects.

Lizard-13 replied:
Nice!, will read the `ObjectsEditorService` in-depth.

Lizard-13 commented:
The polygon editor is up:
![custompolygon](https://user-images.githubusercontent.com/9782647/50552685-24d39b80-0c77-11e9-82da-8472f0a69d5c.gif)

Things to check out:
- Loading vertices from game files don't work, you can try it out, edit a polygon, save, and open the file in a text editor (you should see the vertices are correctly saved, previews work too), reopen the project in GD and in the behavior editor the vertices raw data (a debug field) will show an empty array (I've added it here to avoid crashes:
https://github.com/4ian/GDevelop/blob/61101c9bcfe0debaa82c412c2e8520e27980f885/newIDE/app/src/BehaviorsEditor/Editors/Physics2Editor/index.js#L169-L181
- Discuss and improve the way to render the dialog and other properties
- I've translated the C++ code to check polygon convexity to use it from the IDE, I can switch to create a gd.Polygon2d in the fly if you want, although it would require to create gd.Vector2f's each time.
- There are warnings about mixing ">" and "!==", it's easy to fix adding a parenthesis, but pretiffy doesn't like the idea and removes them.
- I'm not taking theming into account, AFAIK nothing is themable, will check it out later.
- And, of course, I'm a total noob with react so the design could be totally wrong :D

Lizard-13 commented:
Forget the theme, @4ian you've done it so well that everything is already themable
![darktheme](https://user-images.githubusercontent.com/9782647/50560467-d7d7df80-0cdf-11e9-90dc-540b1bef2302.png)


@zatsme I'm using the same image preview than the sprite collision mask editor, so no zoom right now :(

4ian commented:
This is incredible work! ❤️ I'll take a look at this as soon as possible, thanks for listing the things to check out.

> I'm using the same image preview than the sprite collision mask editor, so no zoom right now :(

Yep, zoom should be something that should be added to work with anything showing a preview. I think it's outside of the scope of the extension.

> @4ian you've done it so well that everything is already themable

That looks great! Yeah the goal is that any component that you use for building the UI is already themed. Glad it worked without you doing anything special :)

> the wiki article is growing (http://wiki.compilgames.net/doku.php/gdevelop5/behaviors/physics2), will start adding images and instructions once the editor is completed :)

Great! The wiki article together with this full featured extensions and editor will be super appreciated by all users!

zatsme commented:
Great that the zoom code when added will work for all 😁

4ian commented on newIDE/app/src/BehaviorsEditor/Editors/Physics2Editor/index.js:
```diff
@@ -0,0 +1,565 @@
+          />
+        </Line>
+        <Line>
+          <div style={{ width: '50%' }}>
```
I think you can use two `<Column noMargin>` instead of div with width 50%

4ian commented on newIDE/app/src/BehaviorsEditor/Editors/Physics2Editor/index.js:
```diff
@@ -0,0 +1,565 @@
+              this.forceUpdate();
+            }}
+          >
+            {[
```
nit: the array seems unnecessary ( {[ ... ]} )

4ian commented on newIDE/app/src/BehaviorsEditor/Editors/Physics2Editor/index.js:
```diff
@@ -0,0 +1,565 @@
+          </SelectField>
+        </Line>
+        <Line>
+          <div style={{ width: '33%' }}>
```
Prefer using `<Column noMargin>` rather than div

4ian commented on newIDE/app/src/BehaviorsEditor/Editors/Physics2Editor/index.js:
```diff
@@ -0,0 +1,565 @@
+
+  renderBitProperty(
+    properties: Object,
+    isLayer: boolean,
```
Rather than passing isLayer, it would be a bit more extensible if you pass "propertyName" :)

Lizard-13 replied:
Yeah, don't know why I did it this way... :P

4ian commented:
(don't care about the approve thing, I just wanted to comment ^^)

Lizard-13 commented:
I'm happy you've not said I'm using the properties and states all upside down :)
Don't worry, more than happy to improve it.

4ian commented:
No, most things make sense :) Will have to check more thoroughly because I only did a quick pass so expect more comments but we're going in the right direction, especially once we have the BehaviorsEditorService! :)

4ian commented on newIDE/app/src/BehaviorsEditor/Editors/Physics2Editor/PolygonEditor.js:
```diff
@@ -0,0 +1,101 @@
+import AddCircle from 'material-ui/svg-icons/content/add-circle';
+import Delete from 'material-ui/svg-icons/action/delete';
+
+type Vertex = {
```
Can you use exact object typing here? ( "burger" {| ... |} notation )

4ian replied:
Also do `export type Vertex = ...` to import it in another file.

4ian commented on newIDE/app/src/BehaviorsEditor/Editors/Physics2Editor/ShapePreview.js:
```diff
@@ -0,0 +1,213 @@
+// @flow
+import * as React from 'react';
+
+type Vertex = {
```
Better do `import { type Vertex } from './PolygonEditor'`

4ian commented on newIDE/app/src/BehaviorsEditor/Editors/Physics2Editor/ShapePreview.js:
```diff
@@ -0,0 +1,213 @@
+    const { draggedVertex } = this.state;
+    if (!draggedVertex) return;
+
+    const pointOnScreen = this._svg.createSVGPoint();
```
I recognize this code ;) I think at some point we'll have to factor this

Lizard-13 replied:
Yeah, I see what you did there, coordinate systems between different elements is a pain :/

4ian commented on newIDE/app/src/BehaviorsEditor/Editors/Physics2Editor/index.js:
```diff
@@ -0,0 +1,565 @@
+
+    this.state = {
+      image: '',
+      imageWidth: 9,
```
Any reason for this 9 here? ^^

Lizard-13 replied:
Haha, that's just a very sexy 0 :D

4ian commented on newIDE/app/src/BehaviorsEditor/Editors/Physics2Editor/index.js:
```diff
@@ -0,0 +1,565 @@
+    });
+  };
+
+  _isPolygonConvex(properties: Object) {
```
nitpicking: properties is a `gdMapStringPropertyDescriptor` (even if this map to Object for now)

4ian commented on newIDE/app/src/BehaviorsEditor/Editors/Physics2Editor/index.js:
```diff
@@ -0,0 +1,565 @@
+    );
+  }
+
+  renderNumericProperty(
```
Note that instead of doing a renderXXX method, you could extract this function out of the component, call it NumericProperty, and have an object as argument. Let's call this argument... props :)

`function NumericProperty(props: {|
     properties: Object,
     name: string,
     limit: boolean,
     min: number,
     step: number
  |})`

And boom, you have a component :) Well actually, you need to pass project and behavior and a "onUpdate" callback. Then you can do in your render:

```
<NumericProperty 
     properties={properties}
     name={...}
     limit
     min={...}
     step={...}
/>
```

(btw, this is what is **super great** with React. The cost of abstracting something in a component is literally to move a function... this make the barrier to refactoring and abstracting things ultra low).

It's a bit more readable that using `renderXXX` methods. It also allow to move these extracted component in other files if needed (here it's fine to keep them in a single file, colocating components is a good thing as long as they are not too big).

Lizard-13 commented:
Applied the suggested changes, "expand" in the columns did the trick. Also moved the polygon convexity check to the polygon editor, has a lot more sense to be there, and added the limit to 8 vertices.

4ian commented:
Great, we're moving in the right direction! Do you think comfortable about making the BehaviorsEditorService we discussed and using it to have either a PropertiesEditor or a custom editor? This should be used in `BehaviorsEditor/index.js` if I'm not mistaken, to replace the PropertiesEditor "hardcoded" here: https://github.com/4ian/GDevelop/blob/857c706f4881c2ed474a84f136a33d37b17f93be/newIDE/app/src/BehaviorsEditor/index.js#L180-L194

Lizard-13 commented:
Yeah don't worry, a bit busy right now but I'll get it to work :+1:
Just wondering, to decide if it uses a custom behavior or the default properties editor, should I do the same I'm doing right now (i.e. iterate searching for a `dialog` perperty) ?, or do you have in mind other thing?

4ian commented:
I was thinking of something based on the `type`, like for objects.

In BehaviorsEditorService:
```
    getEditor: (behaviorType: string) => {
      if (!this.components[behaviorType]) {
        return BehaviorPropertiesEditor; // The "default" editor
      }

      // Render the editor specially registered for this behavior
      return this.components[behaviorType].component;
    }.
    components: {
      'MyExtension::MyBehavior': {
        component: Physics2BehaviorEditor,
     }
   }
```

with BehaviorPropertiesEditor "extracted" from `BehaviorsEditor/index.js`. Something like that (indentation not correct and missing things for sure, I just quickly wrote it):

```
export default BehaviorPropertiesEditor = (props: Props) => {
  const {behaviorName, object, project} = this.props;
  const behavior = object.getBehavior(behaviorName);

            const properties = behavior.getProperties(project);
            const propertiesSchema = propertiesMapToSchema(
              properties,
              behavior => behavior.getProperties(project),
              (behavior, name, value) =>
                behavior.updateProperty(name, value, project)
            );

  return (
                  {propertiesSchema.length ? (
                    <PropertiesEditor
                      schema={propertiesSchema}
                      instances={[behavior]}
                    />
                  ) : (
                    <EmptyMessage>
                      There is nothing to configure for this behavior. You can
                      still use events to interact with the object and this
                      behavior.
                    </EmptyMessage>
                  )}
  );
}
```

This way we have the same property editor as before if no editor is registered, and a specific editor otherwise if the type is a known, registered type.

Lizard-13 commented:
Ah! nice, you can get the real behavior type string through the `GetTypeName` function, that simplifies a lot!, no more `dialog` property tricks :D, will do that asap.

4ian commented:
Right, it's `GetTypeName`! 👍 (it's a bit inconsistent to be honest as in objects it's `GetType` - but the idea is exactly the same)

Lizard-13 commented:
I think it's done. A bit worried because everything was very straightforward and worked in the first try without errors, just some flow warnings.

4ian commented:
Failing tests are because you're not up to date with master (+ have a non prettified file). Will ignore this, run Prettier, re-read everything and merge if this looks good! :)

4ian commented on Extensions/Physics2Behavior/physics2runtimebehavior.js:
```diff
@@ -290,7 +307,55 @@ gdjs.Physics2RuntimeBehavior.prototype.createShape = function() {
+      // Overwrite the vertices stored in the buffer
+      var offset = 0;
+      for (var i = 0, len = this.polygon.vertices.length; i < len; i++) {
+        Box2D.HEAPF32[(this._verticesBuffer + offset) >> 2] =
```
Why the >> 2? I also found it on https://github.com/kripken/box2d.js/issues/87 but not sure exactly why it's useful when addressing the memory here.

Lizard-13 replied:
To divide by the size of float (4) and get the correct position, shift automatically deals with the decimal fraction.
It is a headache and would be a lot simpler if the support for `setValue` would not have been removed :(

4ian commented on newIDE/app/src/BehaviorsEditor/Editors/Physics2Editor/index.js:
```diff
@@ -0,0 +1,586 @@
+  constructor(props: Props) {
+    super(props);
+
+    this.resourcesLoader = ResourcesLoader;
```
In fact, I think we can entirely avoid a constructor by doing:

```
      resourcesLoader = ResourcesLoader;
 
      state = {
       image: '',
       imageWidth: 0,
       imageHeight: 0,
     };
```

Lizard-13 replied:
Haha, something I would love to change, readed the constructor was the recommend way when you had default state values, but I trust you :)

4ian commented on newIDE/app/src/BehaviorsEditor/Editors/Physics2Editor/index.js:
```diff
@@ -0,0 +1,586 @@
+          )}
+        </Line>
+        <Line>
+          <SemiControlledTextField // Debug vertices raw content
```
I guess this is to be removed once we've fixed the issue with vertices?

Lizard-13 replied:
Sure, this and the vertices array check/reset.

4ian commented on newIDE/app/src/BehaviorsEditor/index.js:
```diff
@@ -177,20 +168,13 @@ export default class BehaviorsEditor extends Component {
-                    </EmptyMessage>
-                  )}
-                </div>
+                <BehaviorComponent
```
Looks great 👍

4ian commented:
Can you explain me again the issue with the vertices loading? I'll try to reproduce the issue and fix it

4ian commented on newIDE/app/src/BehaviorsEditor/Editors/BehaviorPropertiesEditor.js:
```diff
@@ -0,0 +1,43 @@
+    );
+
+    return (
+      <div style={styles.propertiesContainer}>
```
I've copied this editor almost exactly for the future default editor for objects 👍
Note that we can use `<Column>` instead of the `div` and remove the custom style. This will use the margins from `Column` which are standardized across the app.

Lizard-13 replied:
Will change it then :)

4ian replied:
See https://github.com/4ian/GDevelop/pull/832/files#diff-ed36f79c6072b1dc0fcf031371c586c6R23 :)

Lizard-13 commented:
The vertices array is not loaded properly from the game file (initial game parsing), edit an object with a polygon and save, if you edit the game file you'll notice the vertices are properly saved. Open the game again in GD and the vertices array is converted to a degenerated object instantly.

The weird part is that I've un/serialized the behavior content in a hardcoded string and the serializer loads the vertices correctly.

Lizard-13 commented:
Seems that the only missing points from my original list are the bad vertices loading and the warnings about mixing ">" and "!==" without parenthesis removed by prettify.

4ian commented:
Ok, for the warnings about mixing operators, put `zCrossProduct > 0` in a boolean `zCrossProductIsPositive` and compare against it: `zCrossProductIsPositive !== zProductIsPositive`. That will force us to be a bit more explicit ;)

4ian commented:
Found the culprit for the vertices! Turns out that this was the issue: https://github.com/4ian/GDevelop.js/commit/7feb8d3f1879dbf3f8c8f998c8bb097b5da7e0f4#diff-0b0a8f42de3a19264369288a15d8b56bR133

So in JavaScript, with GDevelop.js, it's more efficient to let JavaScript parse JSON (using JSON.parse) and then traverse this object to create the associated gd.SerializerElement (that can feed into GDevelop.js) instead of using gd.Serializer.fromJSON (which does the parsing using the C++ written JSON serializer) (It's also easier in JavaScript to store a JS object rather than converting everything back and forth to JSON, for example for the undo/redo history.)

So what I did was to add an adhoc "special" function called gd.Serializer.fromJSObject. This is the function doing the step I was talking about "traverse this object to create the associated gd.SerializerElement".

The issue was that I properly fixed earlier gd.Serializer.fromJSON to understand arrays and store them as SerializerElement that are arrays, but I totally forgot to update gd.Serializer.fromJSObject.
I've fixed this and update libGD.js

I was able to save a polygon, open it again, modify it, save it and open it again. Ideally, I will add a test case for this in the tests of GDevelop.js.

Lizard-13 commented:
I've opened a pull request in GDevelop.js, hehe I won!

4ian commented:
Oh sorry I missed the pull request (but I've made sure to be notified of any future one). Good job, you won indeed, here is your medal: 🥇;)
So we should be very close to be ready here. Triple check that everything is working and I'll merge this.

Lizard-13 commented:
Yeah, let me do some save/loads and polygon tests

Lizard-13 commented:
Thanks for the medal!
![lizardinwon](https://user-images.githubusercontent.com/9782647/50708272-806ca300-1042-11e9-8b9a-99d0b48a5bbf.png)

Lizard-13 commented:
Creating and deleting instances (it erases manually allocated memory, so it's very important):
![multipleinstances](https://user-images.githubusercontent.com/9782647/50710680-6d5dd100-104a-11e9-9e27-b76ffd2e6666.gif)

Resizing shape from different origins, note the concave polygon falls back to a box and is not affected by shape scale:
![resizeshape](https://user-images.githubusercontent.com/9782647/50710681-6d5dd100-104a-11e9-883c-c14e36ea8d6d.gif)

Lizard-13 commented:
I've noticed that the polygon editor renders the top-left polygon origin in the same way than from the origin point, not sure if it's possible to fix?, can you access the object origin point from the IDE?

Lizard-13 commented:
Yeah, that would require lot of work, as it has to be abstracted from the object type, for example should the preview render a text object (using the default string)?

A nice future goal, make the image preview not truncated at the image topleft, support zoom, image origin and maybe a coordinates converter. And add an object renderer that can make a render of an object into a texture, and pass this texture and the object origin to the image preview.

4ian commented:
> Yeah, that would require lot of work, as it has to be abstracted from the object type, for example should the preview render a text object (using the default string)?

I think ideally it should work by creating a Pixi renderer (instead of the image) and then rendering the instance (the renderers extending RenderedInstance) like it's done in the scene editor (`ObjectsRenderingService.createNewInstanceRenderer`).
So that you'll get the exact object as rendered on the scene editor (of course you can't resize it or modify the angle, but that's the point). And you get for free the rendering of all object types already done for you :) (and people writing objects only have one renderer to write for the IDE)

4ian commented:
Will merge now :)

4ian commented:
Huge huge thanks @Lizard-13 for working on this! This new editor will be something that will definitely help people take advantage of the new physics engine!!

As next steps, we should I think concentrate on having a few examples and a exhaustive wiki page.
Ideally, we should remove all the "old" examples or rework them to use the Physics2 engine :)

4ian commented:
Though it seemed that I've catched a bug: I opened a demo (physics-joints-demo.json) and tried to set a polygon for `RevoluteBox`, but when lauching the game the vertices are not defined: 
![image](https://user-images.githubusercontent.com/1280130/50724780-a057a180-10ea-11e9-829c-04c6c3ba7686.png)

Lizard-13 commented:
Will take a look right now.

Lizard-13 commented:
Fixed, polygons work perfect, the problem was on not-polygon shapes :/
